### PR TITLE
WIP Add Rename Handle And Change The Copy Operation Logic

### PR DESCRIPTION
**Pull request by fireworkmarks**

### CODE REVIEW

File: api/api-router.go [hunk 1]
```diff
@@ -66,6 +66,9 @@ func RegisterAPIRouter(mux *router.Router, api ObjectAPIHandlers) {
 		// CopyObject
 		bucket.Methods("PUT").Path("/{object:.+}").HeadersRegexp("X-Amz-Copy-Source", ".*?(/).*?").
 			HandlerFunc(api.CopyObjectHandler)
+		// MoveObject
```
Comment (cxt90730) [inline api/api-router.go]: Rename Object

File: api/object-handlers.go [hunk 2]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+	renameSource := r.Header.Get("X-Amz-Copy-Source")
+	// Skip the first element if it is '/', split the rest.
+	if strings.HasPrefix(renameSource, "/") {
+		renameSource = renameSource[1:]
+	}
+	splits := strings.SplitN(renameSource, "/", 2)
+
+	// Save sourceBucket and sourceObject extracted from url Path.
+	var sourceBucketName, sourceObjectName, sourceVersion string
+	if len(splits) == 2 {
```
Comment (cxt90730) [inline api/object-handlers.go]: if length not 2, return err

File: api/object-handlers.go [hunk 3]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		sourceBucketName = splits[0]
+		sourceObjectName = splits[1]
+	}
+	// If source object is empty, reply back error.
+	if sourceObjectName == "" {
```
Comment (cxt90730) [inline api/object-handlers.go]: bucket_name not empty

File: api/object-handlers.go [hunk 4]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		sourceObjectName = splits[0]
+		if !strings.HasPrefix(splits[1], "versionId=") {
+			WriteErrorResponse(w, r, ErrInvalidRenameSource)
+			return
+		}
+		sourceVersion = strings.TrimPrefix(splits[1], "versionId=")
+	}
+
+	// X-Amz-Copy-Source should be URL-encoded
+	sourceBucketName, err = url.QueryUnescape(sourceBucketName)
```
Comment (cxt90730) [inline api/object-handlers.go]: maybe not needed

File: api/object-handlers.go [hunk 5]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		return
+	}
+	sourceObjectName, err = url.QueryUnescape(sourceObjectName)
+	if err != nil {
+		WriteErrorResponse(w, r, ErrInvalidRenameSource)
+		return
+	}
+
+	//Determine if the renamed object is a folder
+	if string(targetObjectName[len(targetObjectName)-1]) == "/" {
```
Comment (cxt90730) [inline api/object-handlers.go]: has suffix

File: api/object-handlers.go [hunk 6]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		helper.ErrorIf(err, "Unable to Folder renaming with :"+targetObjectName)
+		WriteErrorResponse(w, r, ErrInvalidRenameObjectName)
+		return
+	} else if string(sourceObjectName[len(sourceObjectName)-1]) == "/"{
```
Comment (cxt90730) [inline api/object-handlers.go]: move this case

File: api/object-handlers.go [hunk 7]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		return
+	} else {
+		helper.Debugln("sourceBucketName", sourceBucketName, "sourceObjectName", sourceObjectName,
+			"sourceVersion", sourceVersion)
+	}
+
+	//TODO: Object MultiVersion Judge
+	bucketVersion := api.ObjectAPI.GetBucketMultiVersionInfo(sourceBucketName, credential)
+	if bucketVersion == "Enabled" {
+		WriteErrorResponse(w, r, ErrRenameBucket)
```
Comment (cxt90730) [inline api/object-handlers.go]: Rename this err

File: api/object-handlers.go [hunk 8]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		helper.ErrorIf(err, "Unable to Folder renaming with :"+sourceObjectName)
+		WriteErrorResponse(w, r, ErrInvalidRenameObjectName)
+		return
+	} else {
+		helper.Debugln("sourceBucketName", sourceBucketName, "sourceObjectName", sourceObjectName,
+			"sourceVersion", sourceVersion)
+	}
+
+	//TODO: Object MultiVersion Judge
+	bucketVersion := api.ObjectAPI.GetBucketMultiVersionInfo(sourceBucketName, credential)
```
Comment (cxt90730) [inline api/object-handlers.go]: return err

File: api/object-handlers.go [hunk 9]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+	if bucketVersion == "Enabled" {
+		WriteErrorResponse(w, r, ErrRenameBucket)
+		return
+	}else {
+		helper.Debugln("Bucket Multi-version is:",bucketVersion)
+	}
+
```
Comment (cxt90730) [inline api/object-handlers.go]: source bucket must same as target bucket

File: api/object-handlers.go [hunk 10]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+	targetBucketName := vars["bucket"]
+	targetObjectName := vars["object"]
+
+	var credential common.Credential
+	var err error
+	if credential, err = checkRequestAuth(api, r, policy.PutObjectAction, targetBucketName, targetObjectName); err != nil {
+		WriteErrorResponse(w, r, err)
+		return
+	}
+	renameSource := r.Header.Get("X-Amz-Copy-Source")
```
Comment (cxt90730) [inline api/object-handlers.go]: X-Amz-Copy-Source-Key

File: api/object-handlers.go [hunk 11]
```diff
@@ -620,6 +567,136 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+
+	sourceObject, err := api.ObjectAPI.GetObjectInfo(sourceBucketName, sourceObjectName,
+		sourceVersion, credential)
+	if err != nil {
+		WriteErrorResponseWithResource(w, r, err, renameSource)
+		return
+	}
+
+	// Verify before x-amz-copy-source preconditions before continuing with CopyObject.
+	if err = checkObjectPreconditions(w, r, sourceObject); err != nil {
```
Comment (cxt90730) [inline api/object-handlers.go]: remove it

File: api/object-handlers.go [hunk 12]
```diff
@@ -620,6 +567,82 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		return
+	}
+
+	//Determine if the renamed object is a folder
+	 if hasSuffix(sourceObjectName,"/"){
+		WriteErrorResponse(w, r, ErrInvalidRenameSource)
+		return
+	 }
+
+	//TODO: Object MultiVersion Judge
```
Comment (cxt90730) [inline api/object-handlers.go]: supplement the comment

File: storage/object.go [hunk 13]
```diff
@@ -325,6 +325,13 @@ func (yig *YigStorage) GetObjectInfo(bucketName string, objectName string,
 	return
 }
 
+// TODO：Determine here whether it is a multi-version control object or bucket
```
Comment (cxt90730) [inline storage/object.go]: remove it

File: api/object-handlers.go [hunk 14]
```diff
@@ -620,6 +567,83 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 	pipeReader.Close()
 }
 
+// RenameObjectHandler - Rename Object
+// ----------
+//Do not support bucket to enable multiVersion renaming;
+//Folder renaming operation is not supported.
+func (api ObjectAPIHandlers) RenameObjectHandler(w http.ResponseWriter, r *http.Request){
+	helper.Debugln("RenameObjectHandler", "enter")
```
Comment (dahefanteng) [inline api/object-handlers.go]: Debug log should be meaningful

File: meta/client/client.go [hunk 15]
```diff
@@ -16,7 +16,7 @@ type Client interface {
 	GetAllObject(bucketName, objectName, version string) (object []*Object, err error)
 	PutObject(object *Object, tx interface{}) error
 	UpdateAppendObject(object *Object, tx interface{}) error
-	UpdateObjectAttrs(object *Object) error
+	UpdateObjectName(object *Object, sourceObject string) error
```
Comment (dahefanteng) [inline meta/client/client.go]: Why don't u define the function name as "RenameObject" to make it consistent with the func above.

File: meta/types/bucket.go [hunk 16]
```diff
@@ -11,6 +11,12 @@ import (
 	"time"
 )
 
+const(
+	VersionEnabled   = "Enabled"
+	VersionDisabled  = "Disabled"
+	VersionSuspended ="Suspended"
```
Comment (dahefanteng) [inline meta/types/bucket.go]: Leave a blank space before "Suspended",maybe u should use gofmt to format your code before commit.

File: api/object-handlers.go [hunk 17]
```diff
@@ -620,6 +567,83 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		WriteErrorResponse(w, r, ErrInvalidRenameSourceKey)
+		return
+	}
+
+	//Determine if the renamed object is a folder
+	 if hasSuffix(sourceObjectName,"/"){
+		WriteErrorResponse(w, r, ErrInvalidRenameSourceKey)
+		return
+	 }
+
```
Comment (dahefanteng) [inline api/object-handlers.go]: What if the target object and the sourceObject has different prefix?

Comment (fireworkmarks) [reply api/object-handlers.go]: With some improvements

File: meta/types/bucket.go (hunk 16 above)
Comment (fireworkmarks) [reply meta/types/bucket.go]: Fixd it

File: api/object-handlers.go (hunk 17 above)
Comment (fireworkmarks) [reply api/object-handlers.go]: What we are dealing with is just the object, just to take into account the use of windows user folders, so the form of the folder is referenced on the client. In fact, the backend stores the Key-Value form, so it is only necessary to judge the possible error. Logic issues, the rest of the issues are supported by the client.

File: meta/client/client.go (hunk 15 above)
Comment (fireworkmarks) [reply meta/client/client.go]: Have changed

File: api/object-handlers.go (hunk 14 above)
Comment (fireworkmarks) [reply api/object-handlers.go]: As same as other func

**Pull request merged**